portafly/src/components/AppLayout.tsx

```diff
@@ -0,0 +1,172 @@
+} from '@patternfly/react-core'
+import {AppNavExpandable, IAppNavExpandableProps} from './AppNavExpandable'
+import {AppNavGroup, IAppNavGroupProps} from './AppNavGroup'
+import {AppNavItem, IAppNavItemProps} from './AppNavItem'
+
+export interface IAppLayoutContext {
+  setBreadcrumb: (breadcrumb: React.ReactNode) => void
+}
+
+export const AppLayoutContext = React.createContext<IAppLayoutContext>({
+  setBreadcrumb: () => void 0,
+})
+
+export interface IAppLayoutProps
+  extends Pick<PageHeaderProps, 'logo'>,
+    Pick<PageHeaderProps, 'logoProps'>,
+    Pick<PageHeaderProps, 'avatar'>,
+    Pick<PageHeaderProps, 'toolbar'> {
+  navVariant?: 'vertical' | 'horizontal'
+  navItems?: Array<IAppNavItemProps | IAppNavExpandableProps | IAppNavGroupProps | undefined>
+  navGroupsStyle?: 'grouped' | 'expandable'
+  startWithOpenNav?: boolean
+  theme?: 'dark' | 'light'
+  mainContainerId?: string
+}
+
+export const AppLayout: React.FunctionComponent<IAppLayoutProps> = ({
+  logo,
+  logoProps,
+  navVariant = 'horizontal',
+  navItems = [],
+  navGroupsStyle = 'grouped',
+  toolbar,
+  avatar,
+  startWithOpenNav = true,
+  theme = 'dark',
+  mainContainerId = 'main-container',
+  children,
+}) => {
+  const [isNavOpen, setIsNavOpen] = React.useState(startWithOpenNav)
+  const [isMobileView, setIsMobileView] = React.useState(true)
+  const [isNavOpenMobile, setIsNavOpenMobile] = React.useState(false)
+  const [breadcrumb, setBreadcrumb] = React.useState<React.ReactNode | undefined>()
+  const previousBreadcrumb = React.useRef<React.ReactNode | null>()
+
+  const handleSetBreadcrumb = React.useCallback(
+    (newBreadcrumb: React.ReactNode) => {
+      if (previousBreadcrumb.current !== newBreadcrumb) {
+        previousBreadcrumb.current = newBreadcrumb
+        setBreadcrumb(previousBreadcrumb.current)
+      }
+    },
+    [setBreadcrumb, previousBreadcrumb]
+  )
+
+  const onNavToggleMobile = React.useCallback(() => {
+    setIsNavOpenMobile(!isNavOpenMobile)
+  }, [setIsNavOpenMobile, isNavOpenMobile])
+
+  const onNavToggle = React.useCallback(() => {
+    setIsNavOpen(!isNavOpen)
+  }, [setIsNavOpen, isNavOpen])
+
+  const onPageResize = (props: { mobileView: boolean; windowSize: number }) => {
+    setIsMobileView(props.mobileView)
+  }
+
+  React.useEffect(() => {
+    setIsNavOpen(startWithOpenNav)
+  }, [startWithOpenNav, setIsNavOpen])
+
+  const isVertical = navVariant === 'vertical'
+  const variant = isVertical ? NavVariants.default : NavVariants.horizontal
+
+  const Navigation = React.useMemo(
+    () =>
+      navItems.length > 0 ? (
+        <Nav id="nav-primary-simple" theme={theme}>
+          <NavList id="nav-list-simple" variant={variant}>
+            {navItems.map((navItem, idx) => {
+              if (navItem && navItem.hasOwnProperty('items') && isVertical) {
+                return navGroupsStyle === 'expandable' ? (
+                  <AppNavExpandable
+                    {...(navItem as IAppNavExpandableProps)}
```

Comment:
![image](https://user-images.githubusercontent.com/11672286/75364362-e9a11c80-58bb-11ea-8941-bb6932bf752a.png)
9 levels deep! Too much.. We could extract a component for "NavItem" and do:
```html
<NavList id="nav-list-simple" variant={variant}>
  {navItems.map(navItem => <NavItem navItem={navItem} />}
</NavList>
```